### PR TITLE
fix(runtime): resume live sessions from opencode

### PR DESCRIPTION
**Pull request by Maxsky5**
## Summary
- add runtime-level live session discovery to the agent engine port and implement it in the Opencode SDK adapter
- resume matching live Opencode sessions on repo reopen/reload while keeping Beads session persistence limited to durable metadata only
- harden host-side stale run/reset checks against live OpenCode session status and keep runtime route handling aligned with the endpoint abstraction

## Review notes
- Persisted Beads session records no longer drive liveness. The frontend now treats them as durable metadata and derives live status from `listRuntimeSessions()`.
- The runtime abstraction stays at the port boundary: session discovery is exposed as `AgentEnginePort.listRuntimeSessions(...)`, while Opencode-specific HTTP/SDK details remain inside adapters.
- I found and fixed one follow-up gap during review: live-session reconciliation was tracked once per repo, which could miss tasks added later in the same app session. It is now tracked per task.
- I also tightened the host-side OpenCode status probe to use the actual `RuntimeRoute` endpoint instead of assuming `127.0.0.1`.

## Validation
- `bun run typecheck`
- `bun run lint`
- `bun run test`
- `bun run build`
- `bun run check:rust`
- `bun run test:rust`

## Summary by CodeRabbit

* **New Features**
  * Runtime session discovery API and per-runtime status checks to detect live external sessions.

* **Bug Fixes**
  * Filter out stale build runs when external sessions are not live or status endpoints are unreachable.
  * Persisted sessions now hydrate as stopped until live-state reconciliation completes.

* **Improvements**
  * Enhanced session reconciliation: resume/reattach behavior, retry/backoff for bootstrapping, working-directory normalization, and broader test coverage.

### CODE REVIEW

File: apps/desktop/src-tauri/crates/host-application/src/app_service/opencode_session_status.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        return Err(anyhow!(
+            "OpenCode runtime failed to load session status for {working_directory}: HTTP {status_code}{detail_suffix}"
+        ));
+    }
+
+    let body = extract_http_response_body(response.as_str());
+    serde_json::from_str::<OpencodeSessionStatusMap>(body.as_str()).with_context(|| {
+        format!("Failed parsing OpenCode session status response for {working_directory}")
+    })
+}
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply apps/desktop/src-tauri/crates/host-application/src/app_service/opencode_session_status.rs:126]: Test reply path check

[the remark this answers is not in the document]
Comment (Maxsky5) [reply apps/desktop/src-tauri/crates/host-application/src/app_service/opencode_session_status.rs:126]: Not changing this in the current PR. This helper is a narrow, synchronous probe against the host-managed local Opencode runtime, and there is no existing shared Rust HTTP client in this layer to reuse. Adding a new client dependency here would increase surface area more than the code path justifies. I did tighten the helper to honor the RuntimeRoute endpoint directly and to classify unreachable endpoints explicitly.

File: apps/desktop/src-tauri/crates/host-application/src/app_service/runtime_orchestrator.rs [hunk 2]
```diff
@@ -392,16 +404,109 @@ impl AppService {
...
+        }
+
+        let session_cache_key = format!("{}::{}", run.repo_path, run.task_id);
+        if !sessions_by_repo_task.contains_key(session_cache_key.as_str()) {
+            let sessions = self.agent_sessions_list(run.repo_path.as_str(), run.task_id.as_str())?;
+            sessions_by_repo_task.insert(session_cache_key.clone(), sessions);
+        }
+        let sessions = sessions_by_repo_task
+            .get(session_cache_key.as_str())
+            .ok_or_else(|| anyhow!("Missing cached agent sessions for {}", session_cache_key))?;
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply apps/desktop/src-tauri/crates/host-application/src/app_service/runtime_orchestrator.rs:431]: Not taking this refactor in this PR. Using HashMap::entry here would be equivalent behavior-wise, but this thread is style-only and I want to keep the review-fix commit focused on the session/runtime correctness changes.

File: apps/desktop/src-tauri/crates/host-application/src/app_service/runtime_orchestrator.rs [hunk 3]
```diff
@@ -392,16 +404,109 @@ impl AppService {
...
+                        &run.summary.runtime_route,
+                        run.worktree_path.as_str(),
+                    )?
+                }
+            };
+            runtime_statuses_by_worktree.insert(worktree_key.clone(), statuses);
+        }
+        let statuses = runtime_statuses_by_worktree
+            .get(worktree_key.as_str())
+            .ok_or_else(|| anyhow!("Missing cached OpenCode session statuses for {}", worktree_key))?;
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply apps/desktop/src-tauri/crates/host-application/src/app_service/runtime_orchestrator.rs:478]: Not taking this refactor in this PR for the same reason as the adjacent cache comment: it is a style-only cleanup with no behavior change, and I’m keeping this PR scoped to the runtime/session correctness fixes.

File: apps/desktop/src-tauri/crates/host-application/src/app_service/runtime_orchestrator.rs [hunk 4]
```diff
@@ -392,16 +404,109 @@ impl AppService {
...
+        }
+
+        let worktree_key = super::task_workflow::normalize_path_key(run.worktree_path.as_str());
+        if !runtime_statuses_by_worktree.contains_key(worktree_key.as_str()) {
+            let statuses = match run.summary.runtime_kind {
+                AgentRuntimeKind::Opencode => {
+                    super::load_opencode_session_statuses(
+                        &run.summary.runtime_route,
+                        run.worktree_path.as_str(),
+                    )?
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply apps/desktop/src-tauri/crates/host-application/src/app_service/runtime_orchestrator.rs]: Fixed. Dead local session-status endpoints are now treated as stale for stale-run filtering/reset logic, while malformed HTTP or payload responses still fail fast. I also added regression coverage for the unreachable-endpoint case.

File: apps/desktop/src/state/operations/agent-orchestrator/use-agent-orchestrator-operations.ts [hunk 5]
```diff
@@ -292,6 +285,223 @@ export function useAgentOrchestratorOperations({
...
+
+        const taskIdsToReconcile = new Set<string>();
+        for (const input of runtimeConnections.values()) {
+          const runtimeSessions = await agentEngine.listRuntimeSessions(input);
+          if (cancelled) {
+            return;
+          }
+          for (const runtimeSession of runtimeSessions) {
+            if (runtimeSession.status.type !== "busy" && runtimeSession.status.type !== "retry") {
+              continue;
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply apps/desktop/src/state/operations/agent-orchestrator/use-agent-orchestrator-operations.ts]: Fixed. Repo bootstrap now reconciles idle live runtime sessions too, so open-but-idle sessions are reattached and restored as idle instead of staying stopped.

File: apps/desktop/src-tauri/crates/host-application/src/app_service/runtime_orchestrator.rs [hunk 6]
```diff
@@ -302,8 +304,18 @@ impl AppService {
                     true
                 }
             })
-            .map(|run| run.summary.clone())
-            .collect::<Vec<_>>();
+            .filter_map(|run| {
+                match self.should_expose_run(
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply apps/desktop/src-tauri/crates/host-application/src/app_service/runtime_orchestrator.rs:308]: Not changing this in the current PR. The duplication is confined to two focused test modules and is not affecting production behavior. I’d rather keep that cleanup separate from the runtime/session fixes in this branch.

File: apps/desktop/src/state/operations/agent-orchestrator/lifecycle/load-sessions.ts [hunk 7]
```diff
@@ -436,11 +470,135 @@ export const createLoadAgentSessions = ({
...
+          runtimeId: runtimeResolution.runtimeId,
+          runId: runtimeResolution.runId,
+          runtimeEndpoint: runtimeResolution.runtimeEndpoint,
+          workingDirectory: runtimeResolution.runtimeConnection.workingDirectory,
+          status: nextStatus,
+          promptOverrides: repoPromptOverrides,
+          selectedModel: mergeModelSelection(current.selectedModel, selectedModel ?? undefined),
+        }),
+        { persist: false },
+      );
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply apps/desktop/src/state/operations/agent-orchestrator/lifecycle/load-sessions.ts:576]: Fixed. I added the post-resume stale guard so a repo switch during resumeSession no longer attaches listeners or mutates state for the wrong repo. There is a regression test covering that race now.

File: apps/desktop/src/state/operations/agent-orchestrator/use-agent-orchestrator-operations.ts [hunk 8]
```diff
@@ -292,6 +285,223 @@ export function useAgentOrchestratorOperations({
...
+    if (pendingTaskIds.length === 0) {
+      return;
+    }
+
+    for (const taskId of pendingTaskIds) {
+      bootstrappedTasks.add(taskId);
+      void loadAgentSessions(taskId).catch(() => {
+        const currentSet = bootstrappedTasksByRepoRef.current[activeRepo];
+        currentSet?.delete(taskId);
+      });
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply apps/desktop/src/state/operations/agent-orchestrator/use-agent-orchestrator-operations.ts]: Fixed. Background bootstrap/reconcile failures now schedule explicit delayed retries keyed by repo/task instead of silently depending on some future render.

File: apps/desktop/src/state/operations/agent-orchestrator/use-agent-orchestrator-operations.ts [hunk 9]
```diff
@@ -292,6 +285,223 @@ export function useAgentOrchestratorOperations({
...
+            const key = `${input.runtimeKind}::${runtimeSession.externalSessionId}`;
+            const taskIds = persistedTaskIdsByLiveSessionKey.get(key);
+            if (!taskIds) {
+              continue;
+            }
+            for (const taskId of taskIds) {
+              taskIdsToReconcile.add(taskId);
+            }
+          }
+        }
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply apps/desktop/src/state/operations/agent-orchestrator/use-agent-orchestrator-operations.ts:570]: Fixed. The reconciliation pass no longer filters out idle live runtime sessions, so task/session matching now reattaches any live session the runtime reports, not only busy/retry ones.

File: packages/adapters-opencode-sdk/src/opencode-sdk-adapter.ts [hunk 10]
```diff
@@ -67,6 +69,45 @@ import { WORKFLOW_TOOL_CACHE_TTL_MS } from "./types";
...
+          : typeof retryStatus.next === "number"
+            ? retryStatus.next
+            : 0,
+    };
+  }
+
+  return {
+    type: "idle",
+  };
+};
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/adapters-opencode-sdk/src/opencode-sdk-adapter.ts:118]: Fixed. Runtime session status parsing now only defaults to idle when the status entry is genuinely missing. If Opencode returns a malformed or unsupported status object, the adapter throws an actionable error instead of masking the contract drift.

File: apps/desktop/src/state/operations/agent-orchestrator/use-agent-orchestrator-operations.ts [hunk 11]
```diff
@@ -292,6 +290,306 @@ export function useAgentOrchestratorOperations({
...
+          }
+          const { runtimeConnection } = resolveRuntimeRouteConnection(
+            run.runtimeRoute,
+            run.worktreePath,
+          );
+          const key = `${run.runtimeKind}::${runtimeConnection.endpoint}::${runtimeConnection.workingDirectory}`;
+          runtimeConnections.set(key, {
+            runtimeKind: run.runtimeKind,
+            runtimeConnection,
+          });
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply apps/desktop/src/state/operations/agent-orchestrator/use-agent-orchestrator-operations.ts]: Fixed in c0b6b469. Repo reconciliation now deduplicates runtime-wide session scans by `runtimeKind + endpoint`, so multiple worktrees or active runs on the same Opencode endpoint no longer trigger repeated `session.list()` / `session.status()` sweeps. I also added regression coverage in `use-agent-orchestrator-operations.test.tsx` to pin the endpoint-scoped behavior.

File: packages/adapters-opencode-sdk/src/opencode-sdk-adapter.ts [hunk 12]
```diff
@@ -191,6 +241,40 @@ export class OpencodeSdkAdapter
...
+    );
+    const statusesByDirectory = new Map(statusEntries);
+
+    return sessions.map((session) => ({
+      externalSessionId: session.id,
+      title: session.title,
+      workingDirectory: session.directory,
+      startedAt: toIsoFromEpoch(session.time?.created, this.now),
+      status: toRuntimeSessionStatus(statusesByDirectory.get(session.directory)?.[session.id]),
+    }));
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/adapters-opencode-sdk/src/opencode-sdk-adapter.ts]: Fixed in c0b6b469. `listRuntimeSessions()` now validates each directory status payload as a non-null object map before indexing into it, and throws an actionable error if Opencode returns a malformed payload instead of silently treating the session as idle. There is a regression test in `opencode-sdk-adapter.test.ts` for the non-object payload case.

File: apps/desktop/src/state/operations/agent-orchestrator/use-agent-orchestrator-operations.ts [hunk 13]
```diff
@@ -292,6 +293,310 @@ export function useAgentOrchestratorOperations({
...
+          pendingTaskIds.map(
+            async (taskId) =>
+              [
+                taskId,
+                await loadAgentSessionListFromQuery(appQueryClient, activeRepo, taskId, {
+                  forceFresh: true,
+                }),
+              ] as const,
+          ),
+        );
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply apps/desktop/src/state/operations/agent-orchestrator/use-agent-orchestrator-operations.ts]: Fixed in ddc7049c. The repo-level reconcile pass now isolates task failures instead of failing as one batch: both the persisted-session fan-out and the per-task live reconcile stage use per-task settlement, successful tasks still resume immediately, and only the failed task IDs are scheduled for retry. I also added regression coverage for both failure modes in `use-agent-orchestrator-operations.test.tsx`.,workdir:/Users/maxsky5/projects/perso/openducktor.feat-resume-sessions,yield_time_ms:1000,max_output_tokens:4000}},{

File: packages/adapters-opencode-sdk/src/opencode-sdk-adapter.ts [hunk 14]
```diff
@@ -191,6 +250,43 @@ export class OpencodeSdkAdapter
...
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/adapters-opencode-sdk/src/opencode-sdk-adapter.ts]: Fixed in ddc7049c. listRuntimeSessions() now normalizes session directories once and uses the same normalized key for both the per-directory status fetch and the final statusesByDirectory lookup, so whitespace in SDK directory values can no longer make a live session look idle. I added a regression test in packages/adapters-opencode-sdk/src/opencode-sdk-adapter.test.ts for the whitespace case.

File: apps/desktop/src/state/operations/agent-orchestrator/use-agent-orchestrator-operations.test.tsx [hunk 15]
```diff
@@ -1444,4 +1477,491 @@ describe("use-agent-orchestrator-operations", () => {
...
+              endpoint: "http://127.0.0.1:4444",
+            },
+            startedAt: "2026-02-22T08:00:00.000Z",
+            descriptor: OPENCODE_RUNTIME_DESCRIPTOR,
+          },
+        ] as Awaited<ReturnType<typeof host.runtimeList>>;
+      OpencodeSdkAdapter.prototype.listRuntimeSessions = async (input) => {
+        listRuntimeSessionsCalls += 1;
+        scannedEndpoints.push(input.runtimeConnection.endpoint ?? "");
+        return [
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply apps/desktop/src/state/operations/agent-orchestrator/use-agent-orchestrator-operations.test.tsx:1750]: Not taking this suggestion as written. The extra listRuntimeSessions() call in that test is not duplicate endpoint scanning inside the repo-level matcher; it is the separate task-level reconcile load that re-queries runtime state when reattaching the matched session. Tightening the assertion to an unconditional toBe(1) would make the test assert the wrong invariant. I kept the test focused on the real regression it covers: the repo-level pass no longer scans the same runtime endpoint once per worktree.

File: apps/desktop/src/state/operations/agent-orchestrator/use-agent-orchestrator-operations.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (Maxsky5) [reply apps/desktop/src/state/operations/agent-orchestrator/use-agent-orchestrator-operations.ts]: Clarifying the previous reply: the actual fix in ddc7049c is that both task-fan-out stages in the repo-level reconcile path now settle per task instead of failing as one batch. Successful tasks still resume immediately, and only the failed task IDs are retried. Regression coverage for both the persisted-load failure path and the resume failure path is in apps/desktop/src/state/operations/agent-orchestrator/use-agent-orchestrator-operations.test.tsx.

File: packages/adapters-opencode-sdk/src/opencode-sdk-adapter.ts [hunk 16]
```diff
@@ -191,6 +258,50 @@ export class OpencodeSdkAdapter
...
+          ? statusesByDirectory.get(normalizedDirectory)
+          : undefined;
+      return {
+        externalSessionId: session.id,
+        title: session.title,
+        workingDirectory: normalizedDirectory ?? session.directory,
+        startedAt: toIsoFromEpoch(session.time?.created, this.now),
+        status: toRuntimeSessionStatus(directoryStatuses?.[session.id]),
+      };
+    });
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/adapters-opencode-sdk/src/opencode-sdk-adapter.ts:303]: Addressed in e51eab72, but with a stricter fail-fast variant. I did not filter invalid-directory sessions out because that would silently hide a broken Opencode payload. listRuntimeSessions() now requires every listed session to have a valid normalized directory and throws an actionable error when that contract is violated. I added a regression test for the blank-directory case in packages/adapters-opencode-sdk/src/opencode-sdk-adapter.test.ts.
